### PR TITLE
Optimise as_windowed and apply :reflect padding to the whole input

### PR DESCRIPTION
**Pull request by jonatanklosko**
Changes to `indexed_add` to `put_slice`.

Also, instead of applying `:reflect` padding within each incomplete window, we apply it to the input sequence at the beginning. This is what `torch.stft` does and what huggingface/transformers do in pre-processing.

### CODE REVIEW

File: lib/nx_signal.ex [hunk 1]
```diff
@@ -257,7 +257,7 @@ defmodule NxSignal do
 
     as_windowed_parse_non_reflect_opts(
       shape,
-      Keyword.put(opts, :padding, [{div(window_length, 2), div(window_length, 2) - 1}])
+      Keyword.put(opts, :padding, [{div(window_length, 2), div(window_length, 2)}])
```
Comment (jonatanklosko) [inline lib/nx_signal.ex:262]: @polvalente this changes some output shapes (the currently failing tests), but that's what `torch.stft` does. wdyt?

Comment (polvalente) [reply lib/nx_signal.ex:262]: Does it just decrease the number of windows?

Comment (jonatanklosko) [reply lib/nx_signal.ex:262]: The opposite, it adds an extra window, where half of the window is padded.

Comment (polvalente) [reply lib/nx_signal.ex:262]: I think it's ok then

**Review by polvalente — APPROVED** [review]

**Pull request merged**